merkletree.go

```diff
@@ -0,0 +1,344 @@
+				// replace the value
+				currentNodeUL.value = node.value
+				return nil
+			}
+
+			currentNodeKey := m.computePrivateIndex(currentNodeUL.key)
+
+			currentNodeUL.index = currentNodeKey
+
+			direction := getNthBit(currentNodeKey, position)
+			if direction {
+				newInteriorNode.rightChild = currentNodeUL
+			} else {
+				newInteriorNode.leftChild = currentNodeUL
+			}
+			currentNodeUL.level++
+			currentNodeUL.parent = &newInteriorNode
+			if newInteriorNode.parent.leftChild == nodePointer {
+				newInteriorNode.parent.leftChild = &newInteriorNode
+			} else {
+				newInteriorNode.parent.rightChild = &newInteriorNode
+			}
+			nodePointer = &newInteriorNode
+			node.level--
+		case *interiorNode:
+			currentNodeI := nodePointer.(*interiorNode)
+			direction := getNthBit(key, position)
+
+			if direction { // go right
+				currentNodeI.rightHash = nil
+				if currentNodeI.rightChild == nil {
+					currentNodeI.rightChild = node
+					node.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.rightChild
+				}
+			} else { // go left
+				currentNodeI.leftHash = nil
+				if currentNodeI.leftChild == nil {
+					currentNodeI.leftChild = node
+					node.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.leftChild
+				}
+			}
+			position++
+		default:
+			return ErrInvalidTree
+		}
+	}
+	return nil
+}
+
+func (m *MerkleTree) RecomputeHash() {
+	var prefixBits []bool
+	if m.root.leftHash == nil {
+		m.root.leftHash = m.hashNode(
+			append(prefixBits, false), m.root.leftChild)
+	}
+	if m.root.rightHash == nil {
+		m.root.rightHash = m.hashNode(
+			append(prefixBits, true), m.root.rightChild)
+	}
+}
+
+func (m *MerkleTree) hashNode(prefixBits []bool, node interface{}) []byte {
+	if node == nil {
+		// empty node
+		return hashEmptyNode(m, prefixBits)
+	}
+
+	switch node.(type) {
+	case *interiorNode:
+		currentNodeI := node.(*interiorNode)
+		if currentNodeI.leftHash == nil {
+			currentNodeI.leftHash = m.hashNode(
+				append(prefixBits, false), currentNodeI.leftChild)
+		}
+		if currentNodeI.rightHash == nil {
+			currentNodeI.rightHash = m.hashNode(
+				append(prefixBits, true), currentNodeI.rightChild)
+		}
+		return hashInteriorNode(m, currentNodeI)
+	case *userLeafNode:
+		return hashLeafNode(m, node.(*userLeafNode))
+	}
+	return nil
+}
+
+func (m *MerkleTree) digest(ms ...[]byte) []byte {
+	h := m.hash.Hash
+	defer h.Reset()
+	for _, m := range ms {
+		h.Write(m)
+	}
+	return h.Sum(nil)
+}
+
+// Node manipulation functions
+
+var _ MerkleNode = (*userLeafNode)(nil)
+
+func (node *userLeafNode) Value() []byte {
+	return node.value
+}
+
+func (node *interiorNode) serialize() []byte {
+	input := make([]byte, 0, len(node.leftHash)+len(node.rightHash))
+	input = append(input, node.leftHash...)
+	input = append(input, node.rightHash...)
+	return input
+}
+
+func leafNodeCommitment(m *MerkleTree, key string, value []byte) []byte {
```

Comment:
There's a ```userLeafNode``` type, so this should really be declared as a method on that type. Then, once there are one or more node modules, this method should be moved there.